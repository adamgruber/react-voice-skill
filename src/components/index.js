/* eslint import/prefer-default-export: 0 */
import VoiceSkill from './VoiceSkill';

export { VoiceSkill };
