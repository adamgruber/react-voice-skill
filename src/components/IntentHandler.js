import { createElement } from 'react';
import Base from './Base';
import { renderAlexaResponse } from '../renderer';

class IntentHandler extends Base {
  getValidChildren() {
    return ['Response'];
  }

  renderChildren(handlerContext) {
    const firstChild = this.children[0];

    console.log('IntentHandler renderChildren');
    // console.log(this.children);

    // If there is a handler prop, call it with our handlerContext object
    if (typeof this.props.handler === 'function') {
      const handlerResponse = this.props.handler();
      console.log(handlerResponse);

      // Handlers must return a <Response> element which we need to render
      if (typeof handlerResponse !== 'object') {
        throw new Error('IntentHandler handler function must return a single <Response> element.');
      }

      return renderAlexaResponse(handlerResponse, handlerContext);
    }

    // If there are no children, throw an error
    if (!firstChild) {
      throw new Error('IntentHandler must implement `handler` prop when used without children.');
    }

    // If there are multiple children, throw an error
    if (this.children.length > 1) {
      throw new Error('IntentHandler can only wrap a single <Response> child.');
    }

    // If child is not a <Response>, throw an error
    if (firstChild.props.$$type !== 'Response') {
      throw new Error('IntentHandler can only wrap a single <Response> child.');
    }

    // Turn our Response into a React element and render it
    return renderAlexaResponse(
      createElement('RESPONSE', { ...firstChild.props }),
      handlerContext,
    );

    // If there is only one child and it is a string, call `speak`
    // if (this.children.length === 1 && typeof this.children[0] === 'string') {
    //   this.speak(this.children[0]);
    // }
  }

  /**
   * Render method. Gets the handler context from the SDK and
   * extends this class with its properties to make creating responses easier.
   *
   * @param {object} handlerContext Context object passed to the request handler
   */
  render(handlerContext) {
    // Render our children (builds up our response)
    this.renderChildren(handlerContext);
  }
}

export default IntentHandler;
