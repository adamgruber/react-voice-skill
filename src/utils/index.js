import Alexa from 'alexa-sdk';
import { EventEmitter } from 'events';
import VoiceSkill from '../components/VoiceSkill';
import IntentHandler from '../components/IntentHandler';
import Response from '../components/Response';
import Speak from '../components/Speak';
import Card from '../components/Card';

let ROOT_NODE = null;
let RESPONSE_ROOT_NODE = null;

/**
 * Updates the ref to ROOT_NODE / RESPONSE_ROOT_NODE
 *
 * @param {*} rootNode root instance
 */
function getHostContext(rootNode) {
  console.log('getHostContext');
  if (rootNode instanceof EventEmitter) {
    return ROOT_NODE = rootNode;
  }

  return RESPONSE_ROOT_NODE = rootNode;
}

/**
 * Creates an element for a skill
 *
 * @param {string} type Element type
 * @param {Object} props Component props
 * @returns {Object}
 */
function createElement(type, props) {
  console.log(`Create Element: ${type}\n`);
  // console.log(props);
  // console.log(ROOT_NODE);

  const components = {
    ROOT: () => {
      const alexa = Alexa.handler(props.event, props.context);
      alexa.appId = props.appId;
      return alexa;
    },
    RESPONSE_ROOT: () => props.handlerContext,

    VOICESKILL: () => new VoiceSkill(null, { ...props, $$type: 'VoiceSkill' }),
    INTENTHANDLER: () => new IntentHandler(ROOT_NODE, { ...props, $$type: 'IntentHandler' }),

    RESPONSE: () => new Response(RESPONSE_ROOT_NODE, { ...props, $$type: 'Response' }),
    SPEAK: () => new Speak(RESPONSE_ROOT_NODE, { ...props, $$type: 'Speak' }),
    CARD: () => new Card(RESPONSE_ROOT_NODE, { ...props, $$type: 'Card' }),
  };

  const elementCreator = components[type];

  if (elementCreator) {
    return elementCreator();
  } else {
    throw new Error(`Unknown element: ${type}`);
  }
}

export {
  createElement,
  getHostContext,
};
