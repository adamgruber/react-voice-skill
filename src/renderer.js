/* eslint no-param-reassign: 0 */
import ReactFiberReconciler from 'react-reconciler';
import { createElement, getHostContext } from './utils';

const noop = () => {};

const AlexaRenderer = ReactFiberReconciler({
  appendInitialChild(parentInstance, child) {
    console.log('appendInitialChild');
    if (parentInstance.appendChild) {
      parentInstance.appendChild(child);
    } else {
      parentInstance.skill = child;
    }
  },

  createInstance(type, props, internalInstanceHandle) {
    return createElement(type, props, internalInstanceHandle);
  },

  createTextInstance(text) {
    return text;
  },

  finalizeInitialChildren() {
    return false;
  },

  getPublicInstance(instance) {
    return instance;
  },

  prepareForCommit: noop,

  prepareUpdate() {
    return true;
  },

  resetAfterCommit: noop,
  resetTextContent: noop,

  getRootHostContext(rootInstance) {
    return getHostContext(rootInstance);
  },

  getChildHostContext() {
    return {};
  },

  shouldSetTextContent() {
    return false;
  },

  now: noop,

  useSyncScheduling: true,

  mutation: {
    appendChild(parentInstance, child) {
      console.log('mutation appendInitialChild');
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        parentInstance.skill = child;
      }
    },

    appendChildToContainer(parentInstance, child) {
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        switch (child.type) {
          case 'VoiceSkill':
            parentInstance.skill = child;
            break;

          case 'Response':
            parentInstance.res = child;
            break;

          default:
            // ignore
        }
      }
    },

    removeChild: noop,
    removeChildFromContainer: noop,
    insertBefore: noop,
    commitUpdate: noop,
    commitMount: noop,
    commitTextUpdate: noop,
  },
});

/**
 * @param  {ReactElement} element
 * @param {object} options
 * @param {string} options.appId Alexa App ID
 * @param {object} options.event Lambda request event
 * @param {object} options.context Lambda request context
 * @return {object}
 */
const renderToAlexa = (element, { appId, event, context }) => {
  const container = createElement('ROOT', { appId, event, context });
  const node = AlexaRenderer.createContainer(container);
  AlexaRenderer.updateContainer(element, node, null);
  const handlers = container.skill.render();
  console.log(handlers);
  container.registerHandlers(handlers);
  return container.execute();
};

/**
 * @param  {ReactElement} element
 * @param {object} handlerContext ASK SDK Handler Context
 * @return {object}
 */
const renderAlexaResponse = (element, handlerContext) => {
  const container = createElement('RESPONSE_ROOT', { handlerContext });
  const node = AlexaRenderer.createContainer(container);
  AlexaRenderer.updateContainer(element, node, null);

  // Get our response object from rendering components
  const response = container.res.render();
  console.log(response);

  // Parse the response object to build our Alexa response
  Object.keys(response).forEach((type) => {
    const res = response[type];
    switch (type) {
      case 'Card':
        container.response.cardRenderer(res.title, res.content, res.imgSrc);
        break;

      case 'Speak':
        container.response.speak(res.content);
        if (res.listen) {
          container.response.listen(res.listen);
        }
        break;

      default:
        // ignore
    }
  });
  return container.emit(':responseReady');
};

export {
  AlexaRenderer,
  renderToAlexa,
  renderAlexaResponse,
};
