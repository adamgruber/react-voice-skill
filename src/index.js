import { renderToVoiceSkill, renderToAlexa } from './renderer';

const Card = 'CARD';
const IntentHandler = 'INTENTHANDLER';
const Response = 'RESPONSE';
const Speak = 'SPEAK';
const VoiceSkill = 'VOICESKILL';

export {
    Card,
    IntentHandler,
    renderToAlexa,
    renderToVoiceSkill,
    Response,
    Speak,
    VoiceSkill,
};
