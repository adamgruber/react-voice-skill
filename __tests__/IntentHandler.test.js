import { createElement } from 'react';
import IntentHandler from '../src/components/IntentHandler';
import { renderAlexaResponse } from '../src/renderer';

jest.mock('react');
jest.mock('../src/renderer');

describe('IntentHandler', () => {
  let subject;

  beforeEach(() => {
    createElement.mockReturnValue('<Element/>');
    subject = new IntentHandler();
  });

  it('should have base properties', () => {
    expect(subject).toMatchSnapshot();
  });

  describe('getValidChildren', () => {
    it('should return array of valid child types', () => {
      expect(subject.getValidChildren()).toEqual(['Response']);
    });
  });

  describe('render ', () => {
    beforeEach(() => {
      subject.renderChildren = jest.fn();
      subject.render('sample context');
    });

    it('should call renderChildren passing args', () => {
      expect(subject.renderChildren).toHaveBeenCalledWith('sample context');
    });
  });

  describe('render children', () => {
    it('should throw when there are no children', () => {
      expect(() => subject.renderChildren())
        .toThrow('IntentHandler must implement `handler` prop when used without children.');
    });

    it('should throw when there are multiple children', () => {
      subject.appendChild({ type: 'Response' });
      subject.appendChild({ type: 'Response' });
      expect(() => subject.renderChildren())
        .toThrow('IntentHandler can only wrap a single <Response> child.');
    });

    it('should create and render a RESPONSE element', () => {
      const props = { children: [] };
      subject.appendChild({ type: 'Response', props });
      subject.renderChildren('context');
      expect(renderAlexaResponse).toHaveBeenCalledWith('<Element/>', 'context');
    });

    describe('when handler prop is defined', () => {
      beforeEach(() => {
        subject.props.handler = jest.fn();
      });

      describe('and does not return a Response', () => {
        beforeEach(() => {
          subject.props.handler.mockReturnValue({});
        });

        it('should throw', () => {
          expect(() => subject.renderChildren())
            .toThrow('IntentHandler handler function must return a <Response> element.');
        });
      });

      describe('and returns Response', () => {
        const ResponseEl = { type: 'RESPONSE' };

        beforeEach(() => {
          subject.props.handler.mockReturnValue(ResponseEl);
          subject.renderChildren('context');
        });

        it('should render response', () => {
          expect(renderAlexaResponse).toHaveBeenCalledWith(ResponseEl, 'context');
        });
      });
    });
  });
});
