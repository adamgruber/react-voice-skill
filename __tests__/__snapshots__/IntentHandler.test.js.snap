// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntentHandler should have base properties 1`] = `
IntentHandler {
  "children": Array [],
  "props": Object {},
  "root": undefined,
  "type": undefined,
}
`;
